Migrate to OIDC Trusted Publishing

tanjeemh opened the pull request:
**What problem are we solving?**
With OIDC Trusted Publishing, GHA authenticates directly with npm using short-lived identity tokens instead. 
- Eliminates the need to store long-lived npm tokens, improving security.
- Reduces secret management complexity across BitGo repositories, making maintenance simple.
- Improved scalability now that our release pipelines are self-contained within GitHub Actions.
- Necessary for npm compliance with their removal of classic npm-token's.

**Why solve it this way?**
This PR migrates our semantic-release workflow to use OIDC Trusted Publishing for npm releases. Previously, our release pipeline required an `npm-token` secret to authenticate with npm during publishing. This approach has since been discouraged by npm themselves. 
```
npm notice SECURITY NOTICE: 
   Breaking changes starting October 13, 2025. 
   New tokens will be limited to a maximum lifetime of 90 days, and TOTP setup will be disabled. 
   Classic tokens will be revoked in November. 
   Update your CI/CD workflows to avoid disruption. Learn more: https://gh.io/npm-token-changes
```

**Fixes:**
- Our semantic-release-action/typescript reusable workflow no longer enforces `npm-token` as a required secret, allowing us to remove that input so GHA knows to publish via OIDC Trusted Publishing.
- Add `id-token: write` permissions to the workflow to allow for GHA to request an OIDC token at runtime.
- Pinned the workflow the the updated version of `semantic-release-action/typescript/release.yml@v3.1.0`

**Testing:**
From our test/check logs, we can confirm that OIDC is being used instead of an npm token to establish a connection to npm.
```
[semantic-release] › ℹ  Start step "verifyConditions" of plugin "@semantic-release/npm"
[semantic-release] [@semantic-release/npm] › ℹ  Verifying OIDC context for publishing from GitHub Actions
[semantic-release] [@semantic-release/npm] › ℹ  OIDC token exchange with the npm registry succeeded
[semantic-release] › ✔  Completed step "verifyConditions" of plugin "@semantic-release/npm"
```

As confirmation of a successful release, see the beta pre-release tag that was created in the repo and in npmjs: https://www.npmjs.com/package/@bitgo-forks/io-ts/v/2.1.5-beta.1
https://github.com/BitGo/io-ts/actions/runs/19077165991
<img width="400" height="250" alt="image" src="https://github.com/user-attachments/assets/db4b8baf-ff99-416d-884a-6f6c7b6962b8" />



*Note: this PR should not trigger a new official release to `@bitgo-forks/io-ts` *

ericcrosson-bitgo commented on .github/workflows/release.yaml:
```diff
@@ -22,8 +22,6 @@ on:
-    uses: semantic-release-action/typescript/.github/workflows/release.yml@1d40c29e2d500f3bcceeb13f95d26a3a1b571f51 # v3.0.20
-    secrets:
-      npm-token: "FAKE_NPM_TOKEN_FOR_SEMANTIC_RELEASE"
+    uses: semantic-release-action/typescript/.github/workflows/release.yml@v3.1.0 # v3.1.0
```
Since this references a 3rd-party GitHub Action, we need to continue pinning it to a full-length commit SHA, as recommended by GitHub[^1].

[^1]: https://docs.github.com/en/actions/reference/security/secure-use#using-third-party-actions

ericcrosson-bitgo commented:
Also, this is a tiny point, but I suggest placing the "update version" commit before the "remove npm-token input" commit, since that order is technically safe, but if we were to run CI on each commit in the current order, there would be errors due to invalid inputs!

ericcrosson-bitgo commented:
@tanjeemh Great PR description. Can you please include hyperlinks to where we conducted our testing? Workflow run logs will only be accessible for 90 days, so consider screenshotting those if necessary, but we will still be able to see the pass/fail outcome of a workflow run. Also suggest linking to the successfully-created beta tag and pre-release on npmjs.com as evidence of a successful run

Commit:
ci: pin to version v3.1.0

The reason we are using the "ci" prefix for the commit message is because We do not want to actually trigger a new release here, take a look at DX-2123 for more details.
Now that `npm-token` is no longer a required secret, we can check the release logs and see that OIDC Trusted Publishing is being used.
The change to the `npm-token` requirement was made in the pinned SHA version 3.1.0 of semantic-release-action/typescript.

Ticket: DX-2243

ericcrosson-bitgo reviewed (approved):
Thanks @tanjeemh !

ericcrosson-bitgo commented:
Darn, it still created a release. I'll unpublish it since it is immaterial.
